PostEditor Help: add missing formatting shortcuts to the help modal

andfinally opened the pull request:
The post editor help modal is missing some information about formatting shortcuts which appears in the corresponding help modal in wp-admin. This PR adds them to the bottom of the list.

![help-modal](https://user-images.githubusercontent.com/1647564/36965660-3e11122e-2052-11e8-8469-7c4a16216af7.png)

https://github.com/Automattic/wp-calypso/issues/4829

andfinally commented:
Hi @marekhrabe, would you mind having a quick look at this one for me if you get a chance? I'm adding some items to the post editor keyboard shortcuts modal for https://github.com/Automattic/wp-calypso/issues/4829.

andfinally commented:
Great, cheers Marek!

spen commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -135,6 +208,16 @@ class HelpModal extends React.Component {
 				</table>
+				<p>{ spaceFormatText }</p>
+				<table className="wpcom-help__table">
+					{ this.getTableHead() }
```
5 bonus awesome-points awarded for avoiding calling this repeatedly in here. ;)
Looking back through `getTableHead` and the functions it calls they're all ultimately 'static-ish' getters but by assigning it's value to a variable we do avoid extra calls to `translate`.

spen commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -105,26 +173,31 @@ class HelpModal extends React.Component {
 
+		const spaceFormatText = translate( 'Formatting shortcuts, key, then space:' );
+
+		const enterFormatText = translate( 'Formatting shortcuts, key, then Enter:' );
```
Given that both `spaceFormatText` and `enterFormatText` are static, I think inlining them may be preferable. That would show them in their direct context and avoid the (albeit slight) indirection.

spen commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -26,46 +26,114 @@ class HelpModal extends React.Component {
 		let columns = [];
 
-		forEach( row, ( text, key ) => {
+		forEach( row, ( obj, i ) => {
```
`obj` and `i` are ambiguous here, especially given `i` is an abbreviation of `index` which we already have in the parenting scope. I find that scenarios like this act as a clue that something could be named differently to better show intent.

I like the move to change the data structure to be an array of arrays. I think that represents the data in a way that is more true to form. The outer array represents the 'rows' and the inner array represents the 'cells' that make up the columns implicitly. Thinking of it that way, I think `cell` or `rowCell` or something along those lines might be a better name for the iteratee and then a derivative of this could be made for it's index (`cellIndex`). For the sake of extra clarity after introducing a second index I'd be tempted to also change the current `index` to `rowIndex`.

An alternative to fiddling with the index variables would be to just use `cell.key` to build the list-item `key` prop (slightly confusing!).

spen commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -26,46 +26,114 @@ class HelpModal extends React.Component {
 	};
 
 	defaultShortcuts = () => {
+		const { translate } = this.props;
```
Nice move here, much nicer than seeing `this.props.translate` repeatedly :)

andfinally commented:
Thanks Spen, really good points! I'll sort those out.

designsimply commented:
I tested with https://calypso.live/?branch=update/post-editor-help-modal and could see the new keyboard shortcut tips as described. 👍

<img width="1280" alt="screen shot 2018-03-21 at wed mar 21 11 42 44 am" src="https://user-images.githubusercontent.com/1119271/37727374-4107ed10-2cfd-11e8-9537-77d43bd6d387.png">
<sup>Seen at https://calypso.live/post/design5279.wordpress.com using Chrome 65.0.3325.162 on macOS 10.13.3.</sup>

I like the feedback at https://github.com/Automattic/wp-calypso/issues/4829#issuecomment-374640007 but don't think it needs to hold up shipping if you'd like to 🚢 this and pull some things from that comment into another issue. 🙂

andfinally commented:
I've added your excellent suggestions, thanks @spen!

akirk commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -105,36 +124,49 @@ class HelpModal extends React.Component {
+					{ tableHead }
 					<tbody>{ this.additionalShortcuts().map( this.renderRow, this ) }</tbody>
 				</table>
+				<p>{ translate( 'Formatting shortcuts, key, then space:' ) }</p>
```
Could we try to improve this text (this will also make it easier to translate)? For example: `Formatting shortcuts: enter the text, then press Space:` Alternatively `Formatting shortcuts: enter the snippet, then press Space:`.

This of course also requires changing the `tableHead` appropriately, but after all these are not keys but text that needs to be entered.

akirk commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -105,36 +124,49 @@ class HelpModal extends React.Component {
+					{ tableHead }
+					<tbody>{ this.spaceFormatShortcuts().map( this.renderRow, this ) }</tbody>
+				</table>
+				<p>{ translate( 'Formatting shortcuts, key, then Enter:' ) }</p>
```
`Formatting shortcuts: enter the text, then press Enter:`

spen commented:
I'd argue that there is an existing issue with the semantics of this table in terms of the final markup. 
We've got a single table with the headings 'key', 'action', 'key' and 'action', but really what we're representing is 2 columns of a single table with only _two_ headings - 'key' and 'action' - but of course, we want to show these headings in both columns so, for the sake of convenience, we write a single table with 4 columns and spread the information out that way.

Because we're laying out the data across the 2 side-by-side tables we're also having to structure the data itself differently to fit. Ideally we could just list every key-action pairing linearly (in a simple array) but instead, because we have to consider these 2 side-by-side tables the structure of the data has to be more complex.

Fixing that is well outside of the scope if this PR, buuuut, I think it's good to recognise this, because I think this detail has caused some confusion, at least for me!

I think that because of the above quirks I may have been a little ambiguous or just confusing with my comments around arrays & cells - so I'm sorry about that.

I liked your previous approach, taking this:
```js
[
  { 1: t( 'Heading 1' ), 2: t( 'Heading 2' ) },
  ...
]
```

and restructuring in to this:

```js
[
  [
    { key: 1, action: t( 'Heading 1' ) }, 
    { key: 2, action: t( 'Heading 2' ) },
  ],
  ...
]
```

Keeping the `key` & `action` coupled via an object is quite a reasonable approach and allows us to be more explicit in `renderRow`.

I mistakenly called those `{key,action}` objects 'cells', since that's how they appear visually whereas they're not strictly individual cells, cell pairs might be a better description. This is where my point about the un-semantic structuring comes in - it makes it hard to properly define what each thing is.

I think my comment about cells might have implied that the above structure wasn't right and might have lead you towards the current structure (`[ 1, t( 'Heading 1' ), 2, t( 'Heading 2' ) ]`). This array structure could be seen as a sort of tupple in that the ordering of the values implies their meaning (hence having to `% 2` them later to determine whether each value is a `key` or an `action`). In most cases I see the implicitness of tupples or anything tupple-like as being a bit fragile as it's relatively easy to accidentally get the ordering or expected number of values wrong.

I think I'd prefer to roll back to 

```
[
  [
    { key: 1, action: t( 'Heading 1' ) }, 
    { key: 2, action: t( 'Heading 2' ) },
  ],
  ...
]
```

And just define `obj` and `i` better... cell doesn't work in place of obj, maybe just something generic like `shortcut`? `rowIndex` still makes sense though.

That was a bit long winded, so thanks for the patience!

andfinally commented:
Ah right, ta for the detailed explanation! I did wonder if the `% 2` condition would be acceptable, even though it allowed a nice simple data structure. `cellPair` might be a more informative name. I'll revert to the object plan, thanks Spen, and I'll try and address Alex's point about the description.

a8ci18n commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -105,36 +173,57 @@ class HelpModal extends React.Component {
 					context: 'Mac shortcuts',
 				} )
-			: this.props.translate( 'Additional shortcuts, Shift + Alt + key:', {
+			: translate( 'Additional shortcuts. Shift + Alt + key:', {
```
:ok: This change is safe, we'll keep the existing 28 translations.

a8ci18n commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -105,36 +173,57 @@ class HelpModal extends React.Component {
 
 		const additionalText = this.props.macosx
-			? this.props.translate( 'Additional shortcuts, Control + Option + key:', {
+			? translate( 'Additional shortcuts. Control + Option + key:', {
```
:ok: This change is safe, we'll keep the existing 24 translations.

a8ci18n commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -105,36 +173,57 @@ class HelpModal extends React.Component {
-			? this.props.translate( 'Default shortcuts, Command + key:', { context: 'Mac shortcuts' } )
-			: this.props.translate( 'Default shortcuts, Ctrl + key:', { context: 'Windows shortcuts' } );
+			? translate( 'Default shortcuts. Command + key:', { context: 'Mac shortcuts' } )
+			: translate( 'Default shortcuts. Ctrl + key:', { context: 'Windows shortcuts' } );
```
:ok: This change is safe, we'll keep the existing 27 translations.

a8ci18n commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -105,36 +173,57 @@ class HelpModal extends React.Component {
 		const defaultText = this.props.macosx
-			? this.props.translate( 'Default shortcuts, Command + key:', { context: 'Mac shortcuts' } )
-			: this.props.translate( 'Default shortcuts, Ctrl + key:', { context: 'Windows shortcuts' } );
+			? translate( 'Default shortcuts. Command + key:', { context: 'Mac shortcuts' } )
```
:ok: This change is safe, we'll keep the existing 26 translations.

a8ci18n commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -105,36 +173,56 @@ class HelpModal extends React.Component {
 		const defaultText = this.props.macosx
-			? this.props.translate( 'Default shortcuts, Command + key:', { context: 'Mac shortcuts' } )
-			: this.props.translate( 'Default shortcuts, Ctrl + key:', { context: 'Windows shortcuts' } );
+			? translate( 'Default shortcuts. Command + key:', { context: 'Mac shortcuts' } )
```
:ok: This change is safe, we'll keep the existing 26 translations.

andfinally commented:
Hi @spen, I've rolled back to the previous data structure. I changed the property name `key` to `shortcut` - it seems to make more sense now.

@akirk, the wording now reads `Formatting shortcuts. Start a new paragraph with the shortcut and press Space to apply the formatting.` and `Formatting shortcuts. These shortcuts are turned into formatting when you press Enter.`

Testing on IE11 I found a couple of pre-existing flexbox issues, where the dialog's aligned to the left of the viewport, and the text in the `dialog__content` div overflows its container.

These relate to IE11 flexbox bug #17 https://github.com/philipwalton/flexbugs#flexbug-17, (flex items aren't correctly justified when they have `max-width`) and another unlisted bug. 

I've fixed the first by adding `flex: 1 0 100%;` to the `wpcom-help__dialog` card. 

Of all the suggested workarounds for the second bug, the only one that's working is to give the card an explicit percentage height. We might want to think about that more, so I'll raise a separate issue, and also create an issue for Marek's design points as @designsimply suggests.

akirk commented:
Fine with the changes, thank you! For clarity sake this now removes the table header and looks like this (in German, mixed with English is because the translations don't exist yet):

![screen shot 2018-03-26 at 16 18 20](https://user-images.githubusercontent.com/203408/37911813-5d044118-3111-11e8-8b91-2c862e809f61.png)

andfinally commented:
Issue for the IE11 text overflow is https://github.com/Automattic/wp-calypso/issues/23618.

spen commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -105,36 +182,56 @@ class HelpModal extends React.Component {
 					context: 'Windows shortcuts',
 				} );
 
+		const tableHead = this.getTableHead();
```
As promised: 🌟🌟🌟🌟🌟
😉

spen commented on client/components/tinymce/plugins/wpcom-help/help-modal.jsx:
```diff
@@ -26,51 +26,128 @@ class HelpModal extends React.Component {
 		let columns = [];
 
-		forEach( row, ( text, key ) => {
+		forEach( row, cellPair => {
```
This works for me :) `cellPair.shortcut` and `cellPair.action` read well 👍

andfinally commented:
Huzzah! Thanks a lot for your help Spen!
